crypto/cms/cms_sd.c

```diff
@@ -227,6 +227,30 @@ int ossl_cms_SignerIdentifier_cert_cmp(CMS_SignerIdentifier *sid, X509 *cert)
         return -1;
 }
 
+/* Method to map any, incl. provider-implemented PKEY types to OIDs */
+static int cms_generic_sign(CMS_SignerInfo *si, int verify)
+{
+    if (!verify) {
+        int hnid;
+        X509_ALGOR *alg1, *alg2;
+        EVP_PKEY *pkey = si->pkey;
+        const char *typename = EVP_PKEY_get0_type_name(pkey);
+
+        CMS_SignerInfo_get0_algs(si, NULL, NULL, &alg1, &alg2);
+        if (alg1 == NULL || alg1->algorithm == NULL)
+            return -1;
+        hnid = OBJ_obj2nid(alg1->algorithm);
+        if (hnid == NID_undef)
+            return -1;
```

Comment:
This seems quite odd. You go to the effort of confirming that a hash is specified and bailing out if there isn't one - but then complete ignore the hash when trying to find the signature OID below.

Reply:
True. That's a (clearly suboptimal, not truly generic) shortcut that works for providers that don't register digests with their signature algorithms. Let me see whether I can do this even more generically...